Panic in

### Problem statement

Recently, running KIC locally I stumbled across this `panic` when running with `FallbackConfiguration=true` feature flag:

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x102645e44]

goroutine 650 [running]:
github.com/kong/kubernetes-ingress-controller/v3/internal/dataplane/fallback.DefaultCacheGraphProvider.CacheToGraph({}, {{0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, {0x0, ...}, ...})
	/Users/patryk.malek@konghq.com/code_/kubernetes-ingress-controller/internal/dataplane/fallback/cache_to_graph.go:27 +0xb4
github.com/kong/kubernetes-ingress-controller/v3/internal/dataplane/fallback.(*Generator).GenerateExcludingBrokenObjects(_, {{0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, {0x0, ...}, ...}, ...)
	/Users/patryk.malek@konghq.com/code_/kubernetes-ingress-controller/internal/dataplane/fallback/fallback.go:37 +0xd8
github.com/kong/kubernetes-ingress-controller/v3/internal/dataplane.(*KongClient).generateFallbackCache(_, {{0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, {0x0, ...}, ...}, ...)
	/Users/patryk.malek@konghq.com/code_/kubernetes-ingress-controller/internal/dataplane/kong_client.go:658 +0x4f8
github.com/kong/kubernetes-ingress-controller/v3/internal/dataplane.(*KongClient).tryRecoveringWithFallbackConfiguration(_, {_, _}, {{0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, {0x0, ...}, ...}, ...)
	/Users/patryk.malek@konghq.com/code_/kubernetes-ingress-controller/internal/dataplane/kong_client.go:602 +0xc0
github.com/kong/kubernetes-ingress-controller/v3/internal/dataplane.(*KongClient).maybeTryRecoveringFromGatewaysSyncError(_, {_, _}, {{0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}, {0x0, ...}, ...}, ...)
	/Users/patryk.malek@konghq.com/code_/kubernetes-ingress-controller/internal/dataplane/kong_client.go:572 +0x174
github.com/kong/kubernetes-ingress-controller/v3/internal/dataplane.(*KongClient).Update(0x140001e2400, {0x102fd67f0, 0x14001004d20})
	/Users/patryk.malek@konghq.com/code_/kubernetes-ingress-controller/internal/dataplane/kong_client.go:503 +0xcfc
github.com/kong/kubernetes-ingress-controller/v3/internal/dataplane.(*Synchronizer).startUpdateServer(0x1400016f7a0, {0x102fd67f0, 0x14001004d20})
	/Users/patryk.malek@konghq.com/code_/kubernetes-ingress-controller/internal/dataplane/synchronizer.go:183 +0xfc
created by github.com/kong/kubernetes-ingress-controller/v3/internal/dataplane.(*Synchronizer).Start in goroutine 201
	/Users/patryk.malek@konghq.com/code_/kubernetes-ingress-controller/internal/dataplane/synchronizer.go:120 +0x394
```

### Acceptance criteria

- [ ] Using `FallbackConfiguration=true` feature flag does not cause `panic`s.

--- a/internal/dataplane/fallback/cache_to_graph.go
+++ b/internal/dataplane/fallback/cache_to_graph.go
@@ -24,6 +24,9 @@ func (p DefaultCacheGraphProvider) CacheToGraph(c store.CacheStores) (*ConfigGra
 	g := NewConfigGraph()
 
 	for _, s := range c.ListAllStores() {
+		if s == nil {
+			continue
+		}
 		for _, o := range s.List() {
 			obj, ok := o.(client.Object)
 			if !ok {

--- a/internal/dataplane/fallback/cache_to_graph_test.go
+++ b/internal/dataplane/fallback/cache_to_graph_test.go
@@ -47,6 +47,11 @@ func TestDefaultCacheGraphProvider_CacheToGraph(t *testing.T) {
 			expectedAdjacencyMap: map[string][]string{},
 		},
 		{
+			name:                 "cache with stores being nil",
+			cache:                store.CacheStores{},
+			expectedAdjacencyMap: map[string][]string{},
+		},
+		{
 			name: "cache with Ingress and its dependencies",
 			cache: cacheStoresFromObjs(t,
 				&netv1.Ingress{

--- a/internal/dataplane/kong_client.go
+++ b/internal/dataplane/kong_client.go
@@ -598,6 +598,9 @@ func (c *KongClient) tryRecoveringWithFallbackConfiguration(
 	currentCache store.CacheStores,
 	brokenObjects []fallback.ObjectHash,
 ) error {
+	if !currentCache.Available() {
+		return errors.New("failed to generate fallback configuration: cache snapshot not available")
+	}
 	// Generate a fallback cache snapshot.
 	fallbackCache, generatedCacheMetadata, err := c.generateFallbackCache(currentCache, brokenObjects)
 	if err != nil {

--- a/internal/store/cache_stores.go
+++ b/internal/store/cache_stores.go
@@ -60,3 +60,27 @@ func NewCacheStoresFromObjs(objs ...runtime.Object) (CacheStores, error) {
 	}
 	return c, nil
 }
+
+// CacheStoresLockNotInitializedError is returned when the RW lock in the cache stores is nil.
+// It indicates that the CacheStores may not be correctly initialized.
+type CacheStoresLockNotInitializedError struct{}
+
+var _ error = CacheStoresLockNotInitializedError{}
+
+func (e CacheStoresLockNotInitializedError) Error() string {
+	return "lock of cache stores not initialized"
+}
+
+// Available returns whether the cache is correctly initialized and available for loading/storing objects.
+// When the lock or any of the stores in `nil`, it returns false.
+func (c CacheStores) Available() bool {
+	if c.l == nil {
+		return false
+	}
+	for _, store := range c.ListAllStores() {
+		if store == nil {
+			return false
+		}
+	}
+	return true
+}

--- a/internal/store/cache_stores_snapshot.go
+++ b/internal/store/cache_stores_snapshot.go
@@ -15,6 +15,9 @@ import (
 
 // TakeSnapshot takes a snapshot of the CacheStores.
 func (c CacheStores) TakeSnapshot() (CacheStores, error) {
+	if c.l == nil {
+		return CacheStores{}, CacheStoresLockNotInitializedError{}
+	}
 	// Create a fresh CacheStores instance to store the snapshot
 	// in the c.takeSnapshot method. It happens here because it's
 	// not required to be guarded by a lock.
@@ -63,6 +66,9 @@ func (c CacheStores) TakeSnapshotIfChanged(previousSnapshotHash SnapshotHash) (
 	newHash SnapshotHash,
 	err error,
 ) {
+	if c.l == nil {
+		return CacheStores{}, "", CacheStoresLockNotInitializedError{}
+	}
 	// Initialize all variables that don't need to be guarded by a lock.
 	snapshot = NewCacheStores()
 	listOfStores := c.ListAllStores()

